aspnetcore/performance/blazor/webassembly-best-practices.md

```diff
@@ -0,0 +1,154 @@
+
+* `Virtualize` component ([Shared/Virtualize.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Shared/Virtualize.cs)): A component written in C# that implements <xref:Microsoft.AspNetCore.Components.ComponentBase> to render a set of weather data rows based on user scrolling.
+* `FetchData` component ([Pages/FetchData.razor](https://github.com/aspnet/samples/blob/master/samples/aspnetcore/blazor/Virtualization/Pages/FetchData.razor)): Uses the `Virtualize` component to display 25 rows of weather data at a time.
+
+## Avoid JavaScript interop to marshal data
+
+In Blazor WebAssembly, a JavaScript (JS) interop call must traverse the WebAssembly-JS boundary. Serializing and deserializing content across the two contexts creates processing overhead for the app. Frequent JS interop calls often adversely affects performance. To reduce the marshalling of data across the boundary, determine if the app can consolidate many small payloads into a single large payload to avoid the high volume of context switching between WebAssembly and JS.
+
+## Use System.Text.Json
+
+Blazor's JS interop implementation relies on <xref:System.Text.Json>, which is a high-performance JSON serialization library with low memory allocation. Using <xref:System.Text.Json> doesn't result in additional app payload size over adding one or more alternate JSON libraries.
+
+For migration guidance, see [How to migrate from Newtonsoft.Json to System.Text.Json](/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to).
+
+## Use synchronous and unmarshalled JS interop APIs where appropriate
+
+Blazor WebAssembly offers two additional versions of <xref:Microsoft.JSInterop.IJSRuntime> over the single version available to Blazor Server apps:
+
+* <xref:Microsoft.JSInterop.IJSInProcessRuntime> allows invoking JS interop calls synchronously, which has less overhead than the asynchronous versions:
+
+  ```razor
+  @inject IJSRuntime JS
+
+  @code {
+      protected override void OnInitialized()
+      {
+          var jsInProcess = (IJSInProcessRuntime)JS;
+
+          var value = jsInProcess.Invoke<string>("jsInteropCall");
+      }
+  }
+  ```
+
+* <xref:Microsoft.JSInterop.WebAssembly.WebAssemblyJSRuntime> permits unmarshalled JS interop calls:
+
+  ```javascript
+  function jsInteropCall() {
+    return BINDING.js_to_mono_obj("Hello world");
+  }
+  ```
+
+  ```razor
+  @inject IJSRuntime JS
+
+  @code {
+      protected override void OnInitialized()
+      {
+          var jsInProcess = (WebAssemblyJSRuntime)JS;
+
+          var value = jsInProcess.InvokeUnmarshalled<string>("jsInteropCall");
+      }
+  }
+  ```
+
+  > [!WARNING]
+  > While using <xref:Microsoft.JSInterop.WebAssembly.WebAssemblyJSRuntime> has the least overhead of the JS interop approaches, the JavaScript APIs required to interact with these APIs are currently undocumented and subject to breaking changes in future releases.
```

Comment:
I'm wondering whether this should really be included in the docs. If it's undocumented, we shouldn't promote its usage.

Reply:
I'll wait for @pranavkm to let me know before merging. ❓ Pull it? Keep it?

Reply:
Actually ... what I'll do for now is just comment it out. It's too hard to find for the day when it can be live. I'll comment it out in the markup and drop a link to this convo.

Reply:
There are legitimate scenarios where this is useful. It's public API on Blazor's end, and the performance bump is pretty massive.

Reply:
What's the reason for not documenting that API and being afraid that this may not be there?

Reply:
They're mono's apis. They haven't had the time to polish it as yet

Reply:
Got it, thanks!